File: python/deepnoodle/mobius/webhook.py
from __future__ import annotations

import json
import time
import uuid
from collections.abc import Mapping
from dataclasses import dataclass
from typing import Any

import httpx

from .signing import (
    MOBIUS_DELIVERY_ID_HEADER,
    MOBIUS_SECRET_REF_HEADER,
    MOBIUS_SECRET_VERSION_HEADER,
    MOBIUS_SIGNATURE_HEADER,
    MOBIUS_SIGNATURE_VERSION_HEADER,
    MOBIUS_TIMESTAMP_HEADER,
    sign_delivery,
)

WEBHOOK_EVENT_TYPE_HEADER = "X-Mobius-Event-Type"

WEBHOOK_EVENT_RUN_COMPLETED = "run.completed"
WEBHOOK_EVENT_RUN_FAILED = "run.failed"
WEBHOOK_EVENT_PING = "ping"

_SYNTHETIC_WEBHOOK_USER_AGENT = "mobius-sdk-webhook-delivery/1"


@dataclass
class SyntheticWebhookDelivery:
    url: str
    key: bytes
    secret_ref: str
    secret_version: int
    event_type: str
    data: Any
    delivery_id: str | None = None
    timestamp: int | None = None
    http_client: httpx.Client | None = None
    headers: Mapping[str, str] | None = None


def build_synthetic_webhook_payload(event_type: str, data: Any) -> bytes:
    if not event_type:
        raise ValueError("mobius: synthetic webhook event type is required")
    return json.dumps(
        {"type": event_type, "data": data},
        separators=(",", ":"),
    ).encode()


def deliver_synthetic_webhook(delivery: SyntheticWebhookDelivery) -> None:
    if not delivery.url:
        raise ValueError("mobius: synthetic webhook URL is required")
    if not delivery.key:
        raise ValueError("mobius: synthetic webhook signing key is required")
    if not delivery.secret_ref:
        raise ValueError("mobius: synthetic webhook secret ref is required")
    if delivery.secret_version <= 0:
        raise ValueError("mobius: synthetic webhook secret version is required")

    payload = build_synthetic_webhook_payload(delivery.event_type, delivery.data)
    delivery_id = delivery.delivery_id or str(uuid.uuid4())
    timestamp = delivery.timestamp or int(time.time())
    headers = dict(delivery.headers or {})
    headers["Content-Type"] = "application/json"
    headers["User-Agent"] = _SYNTHETIC_WEBHOOK_USER_AGENT
    headers[WEBHOOK_EVENT_TYPE_HEADER] = delivery.event_type
    headers[MOBIUS_SIGNATURE_HEADER] = sign_delivery(
        delivery.key,
        payload,
        delivery_id=delivery_id,
        timestamp=timestamp,
    )
    headers[MOBIUS_SIGNATURE_VERSION_HEADER] = "v1"
    headers[MOBIUS_TIMESTAMP_HEADER] = str(timestamp)
    headers[MOBIUS_DELIVERY_ID_HEADER] = delivery_id
    headers[MOBIUS_SECRET_REF_HEADER] = delivery.secret_ref
    headers[MOBIUS_SECRET_VERSION_HEADER] = str(delivery.secret_version)
    headers["Idempotency-Key"] = delivery_id

    if delivery.http_client is not None:
        resp = delivery.http_client.post(delivery.url, content=payload, headers=headers)
    else:
        with httpx.Client(timeout=60.0) as client:
            resp = client.post(delivery.url, content=payload, headers=headers)
    if resp.status_code < 200 or resp.status_code >= 300:
        raise RuntimeError(
            f"mobius: synthetic webhook returned {resp.status_code}: {resp.text}"
        )

File: typescript/src/webhook.ts
import { randomUUID } from "node:crypto";

import {
  MOBIUS_DELIVERY_ID_HEADER,
  MOBIUS_SECRET_REF_HEADER,
  MOBIUS_SECRET_VERSION_HEADER,
  MOBIUS_SIGNATURE_HEADER,
  MOBIUS_SIGNATURE_VERSION_HEADER,
  MOBIUS_TIMESTAMP_HEADER,
  signDelivery,
} from "./signing.js";

export const WEBHOOK_EVENT_TYPE_HEADER = "X-Mobius-Event-Type";

export type WebhookEventType = "run.completed" | "run.failed" | "ping" | string;

export const WEBHOOK_EVENT_RUN_COMPLETED = "run.completed";
export const WEBHOOK_EVENT_RUN_FAILED = "run.failed";
export const WEBHOOK_EVENT_PING = "ping";

const SYNTHETIC_WEBHOOK_USER_AGENT = "mobius-sdk-webhook-delivery/1";

export interface SyntheticWebhookDelivery {
  url: string;
  key: Uint8Array;
  secretRef: string;
  secretVersion: number;
  deliveryId?: string;
  timestamp?: number;
  eventType: WebhookEventType;
  data: unknown;
  fetch?: typeof globalThis.fetch;
  headers?: HeadersInit;
  signal?: AbortSignal;
}

export function buildSyntheticWebhookPayload(
  eventType: WebhookEventType,
  data: unknown,
): string {
  if (!eventType) {
    throw new Error("mobius: synthetic webhook event type is required");
  }
  return JSON.stringify({ type: eventType, data });
}

export async function deliverSyntheticWebhook(
  delivery: SyntheticWebhookDelivery,
): Promise<void> {
  if (!delivery.url) {
    throw new Error("mobius: synthetic webhook URL is required");
  }
  if (!delivery.key?.byteLength) {
    throw new Error("mobius: synthetic webhook signing key is required");
  }
  if (!delivery.secretRef) {
    throw new Error("mobius: synthetic webhook secret ref is required");
  }
  if (!Number.isInteger(delivery.secretVersion) || delivery.secretVersion <= 0) {
    throw new Error("mobius: synthetic webhook secret version is required");
  }
  const payload = buildSyntheticWebhookPayload(delivery.eventType, delivery.data);
  const deliveryId = resolveDeliveryId(delivery.deliveryId);
  const timestamp = resolveTimestamp(delivery.timestamp);
  const headers = new Headers(delivery.headers);
  headers.set("Content-Type", "application/json");
  headers.set("User-Agent", SYNTHETIC_WEBHOOK_USER_AGENT);
  headers.set(WEBHOOK_EVENT_TYPE_HEADER, delivery.eventType);
  headers.set(
    MOBIUS_SIGNATURE_HEADER,
    signDelivery(delivery.key, Buffer.from(payload), {
      deliveryId,
      timestamp,
    }).signature,
  );
  headers.set(MOBIUS_SIGNATURE_VERSION_HEADER, "v1");
  headers.set(MOBIUS_TIMESTAMP_HEADER, String(timestamp));
  headers.set(MOBIUS_DELIVERY_ID_HEADER, deliveryId);
  headers.set(MOBIUS_SECRET_REF_HEADER, delivery.secretRef);
  headers.set(MOBIUS_SECRET_VERSION_HEADER, String(delivery.secretVersion));
  headers.set("Idempotency-Key", deliveryId);

  const fetchFn = delivery.fetch ?? globalThis.fetch;
  const resp = await fetchFn(delivery.url, {
    method: "POST",
    body: payload,
    headers,
    signal: delivery.signal,
  });
  if (!resp.ok) {
    const text = await resp.text().catch(() => "");
    throw new Error(
      `mobius: synthetic webhook returned ${resp.status}: ${text}`,
    );
  }
}

function resolveDeliveryId(value: string | undefined): string {
  if (value === undefined) {
    return randomUUID();
  }
  if (typeof value !== "string" || value.length === 0) {
    throw new Error("mobius: synthetic webhook deliveryId must be a non-empty string");
  }
  return value;
}

function resolveTimestamp(value: number | undefined): number {
  if (value === undefined) {
    return Math.floor(Date.now() / 1000);
  }
  if (!Number.isInteger(value) || value <= 0) {
    throw new Error("mobius: synthetic webhook timestamp must be a positive integer");
  }
  return value;
}
